### PR TITLE
-migrate liststatus and globstatus operations with tests and code review

### PR DESCRIPTION
**Pull request by github-ccampian**
-migrate liststatus and globstatus operations with tests and code review

### CODE REVIEW

File: src/main/java/org/mule/extension/hdfs/internal/mapping/BeanMapper.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/**
+ * (c) 2003-2017 MuleSoft, Inc. The software in this package is published under the terms of the Commercial Free Software license V.1 a copy of which has been included with this distribution in the LICENSE.md file.
+ */
+package org.mule.extension.hdfs.internal.mapping;
+
+
+/**
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/mapping/BeanMapper.java]: remove signature.

File: src/main/java/org/mule/extension/hdfs/internal/mapping/factory/MapperFactory.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ */
+package org.mule.extension.hdfs.internal.mapping.factory;
+
+import org.apache.tools.ant.filters.StringInputStream;
+import org.dozer.DozerBeanMapper;
+import org.mule.extension.hdfs.internal.mapping.BeanMapper;
+import org.mule.extension.hdfs.internal.mapping.dozer.DozerMapper;
+import org.mule.extension.hdfs.internal.mapping.object.ObjectMap;
+
+/**
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/mapping/factory/MapperFactory.java]: same obs.

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java [hunk 3]
```diff
@@ -40,9 +43,7 @@
      * @return the result from executing the rest of the flow.
      */
     @Throws(HdfsOperationErrorTypeProvider.class)
-    public InputStream readOperation(
-            @Config HdfsConfiguration configuration,
-            @Connection HdfsConnection connection,
+    public InputStream readOperation(@Connection HdfsConnection connection,
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: I don`t think the connection obj is still needed.

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java [hunk 4]
```diff
@@ -40,9 +43,7 @@
...
             String path,
             @Optional(defaultValue = "4096") final int bufferSize)
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: Use @NullSafe instead @Optional

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java [hunk 5]
```diff
@@ -70,9 +71,7 @@ public InputStream readOperation(
      * @param param
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: java doc?

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java [hunk 6]
```diff
@@ -105,14 +104,105 @@ public void write(
      * @return
      */
     @Throws(HdfsOperationErrorTypeProvider.class)
+    public List<FileStatus> listStatus(@Connection HdfsConnection connection, final String path, @Optional final String filter) {
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: same obs.

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java (hunk 4 above)
Comment (github-ccampian) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: nullsafe doesn't have defaul value

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java (hunk 3 above)
Comment (github-ccampian) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: I nedd the connection obj to get the hdfsApiService obj         HdfsAPIService hdfsApiService = serviceFactory.getService(connection);

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java [hunk 7]
```diff
@@ -105,14 +104,105 @@ public void write(
...
+        } catch (Exception e) {
+            throw new ModuleException(e.getMessage(), HdfsErrorType.UNKNOWN, e);
+        }
+    }
+
+    /**
+     * Return all the files that match file pattern and are not checksum files. Results are sorted by their names.
+     * 
+     * @param configuration
+     * @param connection
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: proper java docs or none if not needed.

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java [hunk 8]
```diff
@@ -105,14 +104,105 @@ public void write(
...
+    }
+
+    /**
+     * Make the given file and all non-existent parents into directories. Has the semantics of Unix 'mkdir -p'. Existence of the directory hierarchy is not an error.
+     *
+     * @param path
+     *            the path to create directories for.
+     * @param permission
+     *            the file system permission to use when creating the directories, either in octal or symbolic format (umask).
+     * @throws HDFSConnectorException
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: @throws ModuleException

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java [hunk 9]
```diff
@@ -105,14 +104,105 @@ public void write(
...
+            throw new ModuleException(e.getMessage(), HdfsErrorType.INVALID_REQUEST_DATA, e);
+        } catch (Exception e) {
+            throw new ModuleException(e.getMessage(), HdfsErrorType.UNKNOWN, e);
+        }
+    }
+
+    /**
+     * Delete the file or directory located at the designated path.
+     * 
+     * @param configuration
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: same obs with java doc

Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: remove connection param if not needed.

File: src/main/java/org/mule/extension/hdfs/internal/service/exception/HdfsException.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
+/**
+ * (c) 2003-2017 MuleSoft, Inc. The software in this package is published under the terms of the Commercial Free Software license V.1 a copy of which has been included with this distribution in the LICENSE.md file.
+ */
+package org.mule.extension.hdfs.internal.service.exception;
+
+/**
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/service/exception/HdfsException.java]: signature.

File: src/main/java/org/mule/extension/hdfs/internal/service/impl/FileSystemApiService.java [hunk 11]
```diff
@@ -78,27 +87,39 @@ public void create(String path, String permission, boolean overwrite, int buffer
...
-                    }
-                };
-                return Arrays.asList(fileSystem.listStatus(hdfsPath, pathFilter));
+                PathFilter pathFilter = getPathFilter(hdfsPath, filter);
+                return mapFileStatusFiles(fileSystem.listStatus(hdfsPath, pathFilter));
+            }
+            return mapFileStatusFiles(fileSystem.listStatus(hdfsPath));
+
+        } catch (ConnectException e) {
+            throw new HdfsConnectionException("Something went wrong while sending data to hadoop: " + e.getMessage(), e);
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/service/impl/FileSystemApiService.java]: Check Salesforce for unified message exception handling

File: src/main/java/org/mule/extension/hdfs/internal/service/impl/FileSystemApiService.java [hunk 12]
```diff
@@ -78,27 +87,39 @@ public void create(String path, String permission, boolean overwrite, int buffer
...
+        }
+    }
+
+    @Override
+    public List<FileStatusDTO> globStatus(String pathPattern, String filter)
+            throws InvalidRequestDataException, UnableToRetrieveResponseException, UnableToSendRequestException, HdfsConnectionException {
+
+        try {
+            Path hdfsPath = new Path(pathPattern);
+            if (StringUtils.isNotEmpty(filter)) {
```
Comment (DBCSV) [inline src/main/java/org/mule/extension/hdfs/internal/service/impl/FileSystemApiService.java]: isNotBlank is better so we avoid useless exceptions.

File: src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                .getPath()
+                .toString(), containsString("2013/12/30"));
+    }
+
+    @Test
+    public void testGlobStatusWhenNoFileMatches() throws Exception {
+
+        List<FileStatus> fileStatuses = (List<FileStatus>) flowRunner(Util.FlowNames.GLOB_STATUS_FLOW).withVariable("pathPattern", PARENT_DIRECTORY + INVALID_PATH_PATTERN)
+                .run()
+                .getMessage()
```
Comment (DBCSV) [inline src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java]: .run().getMessage().getPayload().getValue() can be extracted to a common method.

File: src/main/java/org/mule/extension/hdfs/internal/service/impl/FileSystemApiService.java (hunk 12 above)
Comment (github-ccampian) [inline src/main/java/org/mule/extension/hdfs/internal/service/impl/FileSystemApiService.java]: Done

File: src/test/java/org/mule/extension/hdfs/automation/functional/ListStatusTestCases.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+import org.mule.extension.hdfs.util.Util;
+import org.mule.modules.hdfs.automation.functional.BaseTest;
+
+public class ListStatusTestCases extends BaseTest {
+
+    private final String PARENT_DIRECTORY = "hdfs-test-list-status/";
+    private final String FILE_PATH1 = PARENT_DIRECTORY + "hdfs-test-write-file1.txt";
+    private final String FILE_PATH2 = PARENT_DIRECTORY + "hdfs-test-write-file2.txt";
+
+    private final String OWNER = "adriaan";
```
Comment (DBCSV) [inline src/test/java/org/mule/extension/hdfs/automation/functional/ListStatusTestCases.java]: rename this.

File: src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+                .getPayload()
+                .getValue();
+
+        assertThat(fileStatuses, notNullValue());
+        assertThat(fileStatuses, empty());
+        assertThat(fileStatuses.size(), is(0));
```
Comment (DBCSV) [inline src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java]: I think this is redundant.

File: src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+import org.mule.extension.hdfs.util.Util;
+import org.mule.modules.hdfs.automation.functional.BaseTest;
+
+public class GlobStatusTestCases extends BaseTest {
+
+    private final String PARENT_DIRECTORY = "hdfs-test-glob-status/";
+    public static final String INVALID_PATH_PATTERN = "invalidPathPattern";
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
```
Comment (DBCSV) [inline src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java]: is this used?

File: src/main/java/org/mule/extension/hdfs/internal/mapping/BeanMapper.java (hunk 1 above)
Comment (github-ccampian) [inline src/main/java/org/mule/extension/hdfs/internal/mapping/BeanMapper.java]: Done

File: src/main/java/org/mule/extension/hdfs/internal/mapping/factory/MapperFactory.java (hunk 2 above)
Comment (github-ccampian) [inline src/main/java/org/mule/extension/hdfs/internal/mapping/factory/MapperFactory.java]: Done

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java (hunk 5 above)
Comment (github-ccampian) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: Done

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java (hunk 6 above)
Comment (github-ccampian) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: Done

File: src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java (hunk 8 above)
Comment (github-ccampian) [inline src/main/java/org/mule/extension/hdfs/internal/operation/HdfsOperations.java]: fixed

File: src/main/java/org/mule/extension/hdfs/internal/service/exception/HdfsException.java (hunk 10 above)
Comment (github-ccampian) [inline src/main/java/org/mule/extension/hdfs/internal/service/exception/HdfsException.java]: done

File: src/test/java/org/mule/extension/hdfs/automation/functional/ListStatusTestCases.java (hunk 14 above)
Comment (github-ccampian) [inline src/test/java/org/mule/extension/hdfs/automation/functional/ListStatusTestCases.java]: done

File: src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java (hunk 15 above)
Comment (github-ccampian) [inline src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java]: done

File: src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java (hunk 16 above)
Comment (github-ccampian) [inline src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java]: fixed

File: src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java (hunk 13 above)
Comment (github-ccampian) [inline src/test/java/org/mule/extension/hdfs/automation/functional/GlobStatusTestCases.java]: done

**Pull request merged**